### ISSUE TITLE
Feature request - store locale data at row level

### ISSUE THREAD

**Issue by lekoala** [tractorcow/silverstripe-fluent#107]
I was wondering if it would be possible that Fluent stores locale data at row level instead of columns. The reason is that if you have a lot of languages and a lot of fields, things can become wild quite quickly. I'm also wondering this because I'm working on a site that uses the Subsite module and I'd like to be able to set languages per Subsite without having all locales defined in the base table.

Basically, it would mean that for each object where FluentExtension is applied, an extra table is created to store locale date.

I tried to do that by myself a few months ago, but the main problem was how the orm works and basically it's not very obvious how to load fields properly into the object

**Comment by tractorcow** [comment]
I would like to develop an option to allow translated fields to be stored via JSON in a single field, rather than creating additional DB fields for each language and field permutation. That might solve some of the scalability issues without compromising some of the core design decisions of this module.

Using additional tables per locale would run into problems with versioning, but can probably be achieved with lots and lots of additional code, but it also runs into issues where a logical dataobject is now split across multiple tables. It would just about be worth making a separate module to implement this. :)

**Comment by lekoala** [comment]
Well, having one table per locale seems to be a nightmare, but back in SS2 days, I was running a satisfying setup with a "Locale" table with item_id, locale and one column for each translatable fields.

It simply requires a join when doing the query to load locale data, but since the new orm, I didn't spot an easy way to do that, since the fields on the join tables are not loaded (which could be useful on some cases, I'm not sure how the orm is loading the many_many extrafields but it would be a really similar use case).

Maybe it would be possible to define a storage strategy in the module (column, row, json?) and according to that, activate the relevant extension?

**Comment by tractorcow** [comment]
The trouble is that all of those tables also needs to be duplicated or _Live and _versions, which means you have quite a complicated matrix of tables for each logical object.

I made an attempt at some point to implement a proof of concept, but the result is that I ran into complications with versioned tables and decided to try an easier solution.

**Comment by phptek** [comment]
@tractorcow 

> I would like to develop an option to allow translated fields to be stored via JSON in a single field, r

FYI, my `JSONText` (https://github.com/phptek/silverstripe-jsontext) module was designed for 
















_exactly_ this use-case. And, unlike the last time I mooted it, it is also now stable at 1.0.1

**Comment by tractorcow** [comment]
Yay, I love it. That's even more powerful that I thought it could be... you even got it working on mysql too?

**Comment by phptek** [comment]
Yup. Travis is green. Note this is obviously an abstraction. The module does not leverage the DB's own JSON engine and simply uses a DBField subclass for storage and JSON query for querying it.

**Comment by dhensby** [comment]
@lekoala have you seen https://github.com/Internetrix/silverstripe-fluent-extra-table

**Comment by lekoala** [comment]
interesting ! however, by looking at the code, it really shows that dealing with Versioned is not so easy.
keeping everything in a json field might be actually much simpler. The only downside is that the search functionality might get tricky to use properly, particularly if not using mysql 5.7 features.
What do you think? Maybe there is no "best" solution to this, at that the easiest way is to allow the developer to select which strategy fits best his use case : the current way, json storage or extra table. I strongly believe it would make sense to fit all theses strategies under one single module to allow for consistent upgrade and support.

**Comment by phptek** [comment]
@lekoala which Mysql features are you referring to? Note the module makes no use of the underlying RDBMS' JSON support. It's quite simply a blob of JSON as a string stored in a DBField. The power comes with the querying and return type casting.

**Comment by lekoala** [comment]
I'm thinking about these

https://dev.mysql.com/doc/refman/5.7/en/json.html

We could have something like this to select for example a French locale

`SELECT JSON_EXTRACT('{"en": .., "fr": ...}', '$.fr');`

I'm mostly concerned by the ability to filter (like, etc) and order the content based on locale data.

**Comment by phptek** [comment]
An additional and optional `Extension` within the `JSONtext` package could deal to this and allow userland logic to user either `JSON_EXTRACT()` if the module detects that they're using MySQL >=5.7.8 otherwise default to the built-in JSONPath quering that the module provides. For more on JSONPath see: http://goessner.net/articles/JsonPath/

**Comment by tractorcow** [comment]
I've had some reports of issues running fluent with MySQL 5.7. I need to look at building this into our test suite, FYI.

**Comment by robbieaverill** [comment]
We've switched Fluent to store at a row level rather than column level for SilverStripe 4's version, closing
